language/move-lang/src/typing/translate.rs

```diff
@@ -87,30 +87,37 @@ fn main_function_(context: &mut Context, name: FunctionName, f: N::Function) ->
 }
 
 fn check_primitive_main_arg(context: &mut Context, mloc: Loc, ty: &Type) {
+    use BuiltinTypeName_ as BT;
+
     let sp!(loc, ty_) = ty;
-    match (ty_.builtin_name(), ty_) {
-        (Some(_), Type_::Apply(_, _, ss)) => ss
-            .iter()
-            .for_each(|s| check_primitive_main_arg(context, mloc, s)),
-        _ => check_primitive_main_arg_error(
-            context,
-            mloc,
-            *loc,
-            core::error_format(ty, &Subst::empty()),
-        ),
+    if let Some(bt) = ty_.builtin_name() {
+        match bt.value {
+            BT::U8 | BT::U64 | BT::U128 | BT::Bool | BT::Address => return,
+            BT::Vector => {
+                if let Type_::Apply(_, _, ty_args) = ty_ {
+                    if ty_args.len() != 1 {
+                        panic!("ICE vector has {} type arguments", ty_args.len());
+                    }
+                    if let sp!(
+                        _,
+                        Type_::Apply(_, sp!(_, TypeName_::Builtin(sp!(_, BT::U8))), _)
+                    ) = &ty_args[0]
+                    {
+                        return;
+                    }
+                }
+            }
+        }
```

Comment:
You should `subtype` here instead of hand rolling some check.
This could break in all sorts of ways you might not expect. The mostly likely of which
- The type was instantiated and has a type variable
- The type itself was a type variable

However, the current declarations of `subtype`/`join` are not really setup to help you in this way.

Add this function immediately before `subtype` and use it here
```
fn subtype_no_report(
    context: &mut Context,
    pre_lhs: Type,
    pre_rhs: Type,
) -> Result<Type, core::TypingError> {
    let subst = std::mem::replace(&mut context.subst, Subst::empty());
    let lhs = core::ready_tvars(&subst, pre_lhs);
    let rhs = core::ready_tvars(&subst, pre_rhs);
    core::subtype(subst.clone(), &lhs, &rhs).map(|(next_subst, ty)| {
        context.subst = next_subst;
        ty
    })
}
```